package codemodel;

message Position {
  optional string filename = 1;
  optional int32 line = 2;
  optional int32 column = 3;
}

message Scope {
  enum Type {
    MODULE = 0;
    CLASS = 1;
    FUNCTION = 2;
  }

  optional Position declaration_pos = 1;
  optional string name = 2;
  optional Type type = 3;

  repeated Scope child_scope = 4;
  repeated Variable child_variable = 5;
}

message Variable {
  enum Type {
    FUNCTION_ARGUMENT = 0;
    SCOPE_VARIABLE = 1;
    MODULE_REF = 2;
  }

  optional Position declaration_pos = 1;
  optional string name = 2;
  optional Type type = 3;

  repeated string possible_type_id = 4;
}
